_posts/2024-08-04-non-academic-problem.md

```diff
@@ -0,0 +1,134 @@
+---
+layout: post
+title: Non-academic problem (Bridge Finding)
+date: 2024-07-04 12:00:00 +0100
+summary: 'Solving CF 1986F with bridge finding.'
+categories: [codeforces, graph-theory]
+keywords: codeforces, dfs, graph-theory, trees
+tags:
+ - codeforces
+ - dfs
+ - graph-theory
+ - trees
+math: true
+---
+
+In [CF 1986F: Non-academic problem](https://codeforces.com/contest/1986/problem/F) we are given a simple graph. The graph is connected, meaning that every vertex can be reached from every other vertex. Our task is to ruin this property, as in "minimize the number of pairs $1 \leq u \lt v \leq n$, between which there exists a path". To do this, we are allowed to choose and remove a single edge.
+
+## Understanding the Problem
+
+By removing an edge, the graph can be broken into two components. The number of connected pairs we would eliminate is the product of the number of vertices in the components. However, not all edges can split the graph in this way. An edge that connects two, otherwise disconnected subgraphs is called a _bridge_. Bridges in the graph below are marked in red.
+
+![bridge](/assets/posts/2024-08-04-non-academic-problem/bridges.svg)
+
+
+## Solution
+
+Notice that an edge is a bridge if and only if it does not belong to any cycle. If it is part of a cycle, then removing it does not break the graph because all paths that pass through the edge can be rerouted along the cycle. If it is not in any cycle, then there is no path that connects its two endpoints while avoiding the edge, and therefore removing it makes those two points disconnected.
+
+Cycles in a graph can be found via depth-first search. We can extend this algorithm to enumerate bridges. The DFS tree has the special property, that for every chord $(u, v)$, $v$ is either an ancestor, or a descendant of $u$. Since every cycle contains at least one chord, if an edge $(u, v)$ is contained in a cycle, then by searching $v$'s subtree, a chord that connects one of the descendants of $v$ to one of its ancestors can be found.
```

Comment:
@leventegego

> chord

You mean non-tree edge?

I've only heard the phrase chord in the context of [chordal graphs](https://en.wikipedia.org/wiki/Chordal_graph) before, where it means 'an edge that is not part of a cycle but connects two vertices of a cycle'.

> The DFS tree has the special property, that for every chord $(u, v)$, $v$ is either an ancestor, or a descendant of $u$.

The DFS edge categories in English:
- faél = tree edge
- visszaél = back edge
- előreél = forward edge
- keresztél = cross edge

So you're saying $(u,v)$ is not a cross edge.

DFS in the undirected case only results in tree edges and back edges. A forward edge in a directed graph would be explored from the descendant if we ignore the direction of the edge, so it would become a back edge. (And a cross edge would become a tree edge.)

So if $(u,v)$ is a back edge, $v$ is the ancestor.